lib/pushmi_pullyu/aip/downloader.rb

```diff
@@ -0,0 +1,214 @@
+              else
+                "#{@noid}: #{local_path} -- not_found"
+              end
+    PushmiPullyu::Logging.log_aip_activity(@noid, message)
+  end
+
+  ### Directories
+
+  def aip_dirs
+    @aip_dirs ||= OpenStruct.new(
+      objects: 'objects',
+      metadata: 'objects/metadata',
+      logs: 'logs',
+      thumbnails: 'thumbnails'
+    )
+  end
+
+  def make_directories
+    clean_directories
+    PushmiPullyu.logger.debug("#{@noid}: Creating directories ...")
+    aip_dirs.to_h.values.each do |path|
+      FileUtils.mkdir_p(full_local_path(path))
+    end
+    PushmiPullyu.logger.debug("#{@noid}: Creating directories done")
+  end
+
+  def clean_directories
+    return unless File.exist?(base_path)
+    PushmiPullyu.logger.debug("#{@noid}: Nuking directories ...")
+    FileUtils.rm_rf(base_path)
+  end
+
+  def base_path
+    PushmiPullyu::AIP.aip_directory(@noid)
+  end
+
+  def data_path
+    "#{base_path}/data"
+  end
+
+  def full_local_path(path)
+    "#{data_path}/#{path}"
+  end
+
+  ### Files
+
+  def aip_paths
+    @aip_paths ||= OpenStruct.new(
+      main_object: OpenStruct.new(
+        remote: nil, # Base path
+        local: "#{@aip_dirs.metadata}/object_metadata.n3",
+        optional: false
+      ),
+      fixity: OpenStruct.new(
+        remote: '/content/fcr:fixity',
+        local: "#{@aip_dirs.logs}/content_fixity_report.n3",
+        optional: false
+      ),
+      content_datastream_metadata: OpenStruct.new(
+        remote: '/content/fcr:metadata',
+        local: "#{@aip_dirs.metadata}/content_fcr_metadata.n3",
+        optional: false
+      ),
+      versions: OpenStruct.new(
+        remote: '/content/fcr:versions',
+        local: "#{@aip_dirs.metadata}/content_versions.n3",
+        optional: false
+      ),
+      content: OpenStruct.new(
+        remote: '/content',
+        local: nil, # Filename derived from metadata
+        optional: false
+      ),
+
+      # Optional downloads
+      thumbnail: OpenStruct.new(
+        remote: '/thumbnail',
+        local: "#{@aip_dirs.thumbnails}/thumbnail",
+        optional: true
+      ),
+      characterization: OpenStruct.new(
+        remote: '/characterization',
+        local: "#{@aip_dirs.logs}/content_characterization.n3",
+        optional: true
+      ),
+      fedora3foxml: OpenStruct.new(
+        remote: '/fedora3foxml',
+        local: "#{@aip_dirs.metadata}/fedora3foxml.xml",
+        optional: true
+      ),
+      fedora3foxml_metadata: OpenStruct.new(
+        remote: '/fedora3foxml/fcr:metadata',
+        local: "#{@aip_dirs.metadata}/fedora3foxml.n3",
+        optional: true
+      )
+    ).freeze
+  end
+
+  def content_filename
+    return @content_filename unless @content_filename.nil?
+
+    # Extract filename from main object metadata
+    graph = RDF::Graph.load(full_local_path(aip_paths.main_object.local))
+    graph.each_statement do |statement|
+      if statement.predicate == FILENAME_PREDICATE
```

Comment:
Would it be better/faster to try using [`RDF::Graph#query`](https://github.com/ruby-rdf/rdf/blob/develop/examples/query_bench.rb#L13), or is iteration better? I haven't tried it but it seems like iteration could be slow on a large RDF graph.